diff -ruN glusterfs-11.2/api/examples/Makefile.am glusterfs-11.2-mod/api/examples/Makefile.am
--- glusterfs-11.2/api/examples/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/api/examples/Makefile.am	2025-07-02 12:31:15.002167956 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 # The bits needed for glfsxmp
 EXTRA_PROGRAMS = glfsxmp
 glfsxmp_SOURCES = glfsxmp.c
diff -ruN glusterfs-11.2/contrib/fuse-util/Makefile.am glusterfs-11.2-mod/contrib/fuse-util/Makefile.am
--- glusterfs-11.2/contrib/fuse-util/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/contrib/fuse-util/Makefile.am	2025-07-02 12:31:32.290546271 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 bin_PROGRAMS = fusermount-glusterfs
 
 fusermount_glusterfs_SOURCES = fusermount.c mount_util.c $(CONTRIBDIR)/fuse-lib/mount-common.c
diff -ruN glusterfs-11.2/libglusterfs/src/Makefile.am glusterfs-11.2-mod/libglusterfs/src/Makefile.am
--- glusterfs-11.2/libglusterfs/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/libglusterfs/src/Makefile.am	2025-07-02 12:26:47.776161392 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 noinst_PYTHON = generator.py gen-defaults.py $(top_srcdir)/events/eventskeygen.py
 
 libglusterfs_la_CFLAGS = $(GF_CFLAGS) $(GF_DARWIN_LIBGLUSTERFS_CFLAGS) \
diff -ruN glusterfs-11.2/xlators/cluster/afr/src/Makefile.am glusterfs-11.2-mod/xlators/cluster/afr/src/Makefile.am
--- glusterfs-11.2/xlators/cluster/afr/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/cluster/afr/src/Makefile.am	2025-07-02 12:26:21.567940962 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 xlator_LTLIBRARIES = afr.la
 xlatordir = $(libdir)/glusterfs/$(PACKAGE_VERSION)/xlator/cluster
 
diff -ruN glusterfs-11.2/xlators/cluster/dht/src/Makefile.am glusterfs-11.2-mod/xlators/cluster/dht/src/Makefile.am
--- glusterfs-11.2/xlators/cluster/dht/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/cluster/dht/src/Makefile.am	2025-07-02 12:26:07.943843179 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 xlator_LTLIBRARIES = dht.la nufa.la switch.la
 
 AM_CFLAGS = -Wall $(GF_CFLAGS)
diff -ruN glusterfs-11.2/xlators/cluster/ec/src/Makefile.am glusterfs-11.2-mod/xlators/cluster/ec/src/Makefile.am
--- glusterfs-11.2/xlators/cluster/ec/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/cluster/ec/src/Makefile.am	2025-07-02 12:25:45.287655225 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 xlator_LTLIBRARIES = ec.la
 xlatordir = $(libdir)/glusterfs/$(PACKAGE_VERSION)/xlator/cluster
 
diff -ruN glusterfs-11.2/xlators/features/changelog/lib/src/Makefile.am glusterfs-11.2-mod/xlators/features/changelog/lib/src/Makefile.am
--- glusterfs-11.2/xlators/features/changelog/lib/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/features/changelog/lib/src/Makefile.am	2025-07-02 12:25:15.015401482 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 libgfchangelog_la_CFLAGS = -Wall $(GF_CFLAGS) $(GF_DARWIN_LIBGLUSTERFS_CFLAGS) \
 	-DDATADIR=\"$(localstatedir)\"
 
diff -ruN glusterfs-11.2/xlators/features/cloudsync/src/cloudsync-plugins/src/cloudsyncs3/src/Makefile.am glusterfs-11.2-mod/xlators/features/cloudsync/src/cloudsync-plugins/src/cloudsyncs3/src/Makefile.am
--- glusterfs-11.2/xlators/features/cloudsync/src/cloudsync-plugins/src/cloudsyncs3/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/features/cloudsync/src/cloudsync-plugins/src/cloudsyncs3/src/Makefile.am	2025-07-02 12:24:16.783001896 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 csp_LTLIBRARIES = cloudsyncs3.la
 cspdir = $(libdir)/glusterfs/$(PACKAGE_VERSION)/cloudsync-plugins
 
diff -ruN glusterfs-11.2/xlators/features/cloudsync/src/cloudsync-plugins/src/cvlt/src/Makefile.am glusterfs-11.2-mod/xlators/features/cloudsync/src/cloudsync-plugins/src/cvlt/src/Makefile.am
--- glusterfs-11.2/xlators/features/cloudsync/src/cloudsync-plugins/src/cvlt/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/features/cloudsync/src/cloudsync-plugins/src/cvlt/src/Makefile.am	2025-07-02 12:24:00.726984102 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 csp_LTLIBRARIES = cloudsynccvlt.la
 cspdir = $(libdir)/glusterfs/$(PACKAGE_VERSION)/cloudsync-plugins
 
diff -ruN glusterfs-11.2/xlators/features/cloudsync/src/Makefile.am glusterfs-11.2-mod/xlators/features/cloudsync/src/Makefile.am
--- glusterfs-11.2/xlators/features/cloudsync/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/features/cloudsync/src/Makefile.am	2025-07-02 12:24:45.054966491 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 SUBDIRS = cloudsync-plugins
 
 xlator_LTLIBRARIES = cloudsync.la
diff -ruN glusterfs-11.2/xlators/features/thin-arbiter/src/Makefile.am glusterfs-11.2-mod/xlators/features/thin-arbiter/src/Makefile.am
--- glusterfs-11.2/xlators/features/thin-arbiter/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/features/thin-arbiter/src/Makefile.am	2025-07-02 12:22:10.366019070 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 xlator_LTLIBRARIES = thin-arbiter.la
 
 xlatordir = $(libdir)/glusterfs/$(PACKAGE_VERSION)/xlator/features
diff -ruN glusterfs-11.2/xlators/features/utime/src/Makefile.am glusterfs-11.2-mod/xlators/features/utime/src/Makefile.am
--- glusterfs-11.2/xlators/features/utime/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/features/utime/src/Makefile.am	2025-07-02 12:21:54.421923762 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 xlator_LTLIBRARIES = utime.la
 xlatordir = $(libdir)/glusterfs/$(PACKAGE_VERSION)/xlator/features
 
diff -ruN glusterfs-11.2/xlators/mgmt/glusterd/src/Makefile.am glusterfs-11.2-mod/xlators/mgmt/glusterd/src/Makefile.am
--- glusterfs-11.2/xlators/mgmt/glusterd/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/mgmt/glusterd/src/Makefile.am	2025-07-02 12:20:49.141466838 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 if WITH_SERVER
 xlator_LTLIBRARIES = glusterd.la
 endif
diff -ruN glusterfs-11.2/xlators/mount/fuse/src/Makefile.am glusterfs-11.2-mod/xlators/mount/fuse/src/Makefile.am
--- glusterfs-11.2/xlators/mount/fuse/src/Makefile.am	2025-06-10 16:27:14.000000000 +0300
+++ glusterfs-11.2-mod/xlators/mount/fuse/src/Makefile.am	2025-07-02 12:19:50.029047303 +0300
@@ -1,3 +1,5 @@
+AUTOMAKE_OPTIONS = subdir-objects
+
 noinst_HEADERS_linux = $(CONTRIBDIR)/fuse-include/fuse_kernel.h\
 	$(CONTRIBDIR)/fuse-include/mount_util.h\
 	$(CONTRIBDIR)/fuse-lib/mount-gluster-compat.h
